[SPARK-16501] [MESOS] Allow providing Mesos principal & secret via files

rvesse opened the pull request:
## What changes were proposed in this pull request?

This commit modifies the Mesos submission client to allow the principal
and secret to be provided indirectly via files.  The path to these files
can be specified either via Spark configuration or via environment
variable.

Assuming these files are appropriately protected by FS/OS permissions
this means we don't ever leak the actual values in process info like ps

Environment variable specification is useful because it allows you to
interpolate the location of this file when using per-user Mesos
credentials.

For some background as to why we have taken this approach I will briefly describe our set up.  On our systems we provide each authorised user account with their own Mesos credentials to provide certain security and audit guarantees to our customers. These credentials are managed by a central Secret management service. In our `spark-env.sh` we determine the appropriate secret and principal files to use depending on the user who is invoking Spark hence the need to inject these via environment variables as well as by configuration properties. So we set these environment variables appropriately and our Spark read in the contents of those files to authenticate itself with Mesos.

## How was this patch tested?

This is functionality we have been using it in production across multiple customer sites for some time. This has been in the field for around 18 months with no reported issues. These changes have been sufficient to meet our customer security and audit requirements.

We have been building and deploying custom builds of Apache Spark with various minor tweaks like this which we are now looking to contribute back into the community in order that we can rely upon stock Apache Spark builds and stop maintaining our own internal fork.

rvesse committed:
Allow providing Mesos principal & secret via files (SPARK-16501)

This commit modifies the Mesos submission client to allow the principal
and secret to be provided indirectly via files.  The path to these files
can be specified either via Spark configuration or via environment
variable.

Assuming these files are appropriately protected by FS/OS permissions
this means we don't ever leak the actual values in process info like ps

Environment variable specification is useful because it allows you to
interpolate the location of this file when using per-user Mesos
credentials.

AmplabJenkins commented:
Can one of the admins verify this patch?

felixcheung commented:
is putting secrets as plain text files a good practice..?

rvesse commented:
@felixcheung Well currently you are required to provide them in clear text either via `--conf` arguments or in your `spark-defaults.conf` which was unacceptable to our security conscious users. The former is trivially accessible via `ps` output and the latter tends to be a world readable file so both methods are insecure.

Providing some indirection by only having to provide Spark with a path to the file that contains the secrets is an improvement. As noted in my description if these files are properly protected this prevents other users from discovering/exposing those secrets via `ps`, `history` or other means.

rvesse commented:
CC @ArtRand @vanzin I would appreciate your reviews as and when you have time

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala:
```diff
@@ -80,10 +80,27 @@ trait MesosSchedulerUtils extends Logging {
     }
     fwInfoBuilder.setHostname(Option(conf.getenv("SPARK_PUBLIC_DNS")).getOrElse(
       conf.get(DRIVER_HOST_ADDRESS)))
+    conf.getOption("spark.mesos.principal.file")
```
Why is it necessary to read the principal from a file? It's not a really sensitive parameter, and can be provided with `--conf` with no issues.

ArtRand replied:
I agree. I don't think it's necessary to put the principal in a file, just the secret.

rvesse replied:
We have customers who operate secure multi-tenant environments who consider even leaking principals of users from other tenants to be a security issue i.e. they want to minimise what users from one tenant can learn about users from another

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala:
```diff
@@ -80,10 +80,27 @@ trait MesosSchedulerUtils extends Logging {
       credBuilder.setPrincipal(principal)
     }
+    conf.getOption("spark.mesos.secret.file")
+      .orElse(Option(conf.getenv("SPARK_MESOS_SECRET_FILE"))
```
Why not provide the secret directly in the environment, instead of indirectly through a file?

That is more secure than files in certain situations, so it would be nice to at least have that option.

ArtRand replied:
Environment-variable secrets (even using the `secrets` primitive up to Mesos 1.4) will be available at `/proc/<pid>/environ` so file-based is better more secure. I'm tempted to say having an insecure _option_ is worse than having less flexibility.

vanzin replied:
You can't read `/proc/<pid>/environ` from other users' processes.

```
$ cat /proc/1/environ
cat: /proc/1/environ: Permission denied
```

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala:
```diff
@@ -80,10 +80,27 @@ trait MesosSchedulerUtils extends Logging {
       fwInfoBuilder.setPrincipal(principal)
       credBuilder.setPrincipal(principal)
     }
+    conf.getOption("spark.mesos.secret.file")
```
This should be merged with `getOption("spark.mesos.secret")` below.

```
conf.getOption("spark.mesos.secret")
  .orElse(/* Read optional file-based config into string */)
```

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala:
```diff
@@ -80,10 +80,27 @@ trait MesosSchedulerUtils extends Logging {
+    conf.getOption("spark.mesos.secret.file")
+      .orElse(Option(conf.getenv("SPARK_MESOS_SECRET_FILE"))
+      .foreach { secretFile =>
+       val file = io.Source.fromFile(secretFile)
```
You should use `Files.toString()` (from Guava) here (simpler and also handles errors properly).

The whole block is also indented incorrectly.

vanzin reviewed (commented):
You'd probably also need to update the docs.

You should also follow the usual PR title format (described in https://spark.apache.org/contributing.html).

ArtRand commented:
Hello @rvesse thanks for this patch! With Mesos having [reference secrets](https://github.com/apache/mesos/blob/master/include/mesos/mesos.proto#L2596) now I think providing this as a file is a natural evolution.

@vanzin, @rvesse , Why not also allow for `spark.authenticate.secret` to be a file? We have an [patch](https://github.com/mesosphere/spark/commit/a897c1ca07e67935b8ae80cfc1d41370fe8ef60b) at Mesosphere that allows this. It can be in a different JIRA/PR just want an opinion.

rvesse committed:
[SPARK-16501] Improve how Mesos Principal/Secret are specified

Improve implementation based upon PR feedback, simplifies the code and
refactors into own method for unit testing purposes.

Unit tests covering a variety of different combinations of configuration
, environment and file parameters.

rvesse commented:
@vanzin @ArtRand Thanks for the initial reviews, I have refactored the patch based on comments so should hopefully be in a better state now.  I was able to also add some unit test coverage for this functionality.

I will look at also updating documentation appropriately which will need to be a separate PR to the apache/spark-website repository

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

vanzin commented:
> updating documentation appropriately which will need to be a separate PR

Actually the documentation is in `docs/running-on-mesos.md`.

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala:
```diff
@@ -71,40 +74,62 @@ trait MesosSchedulerUtils extends Logging {
+      ).foreach { principal =>
+        fwInfoBuilder.setPrincipal(principal)
+        credBuilder.setPrincipal(principal)
+    }
```
nit: indent extra level

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala:
```diff
@@ -71,40 +74,62 @@ trait MesosSchedulerUtils extends Logging {
+           Files.toString(new File(secretFile), Charset.forName("UTF-8"))
+         }
+      ).foreach { secret =>
+      credBuilder.setSecret(secret)
```
nit: indendation is wrong

vanzin commented on resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtilsSuite.scala:
```diff
@@ -237,4 +242,88 @@ class MesosSchedulerUtilsSuite extends SparkFunSuite with Matchers with MockitoS
+    val conf = new SparkConf()
+    conf.set("spark.mesos.secret", "my-secret")
+    
+    an [SparkException] should be thrownBy utils.buildCredentials(conf, FrameworkInfo.newBuilder())
```
We normally use a different approach:

```
intercept[SparkException] {
  // code
}
```

vanzin reviewed (commented):
I still think env variables are safer than files in the general case. At worst they're the same (unless the process leaks the environment to other users - `/proc` doesn't), at best, the env variables are not written to any unencrypted storage.

So I think that they should at least be an option here if you're adding this feature.

rvesse committed:
[SPARK-16501] Improve Mesos cred specification

Allow specifying both directly via environment variables, improve unit
test coverage and add documentation

rvesse commented:
@vanzin Thanks for the detailed review. I have now added the ability to also specify the credentials directly via environment variables. I have added additional unit test coverage and a new section to the documentation covering these improvements.

ArtRand commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala:
```diff
@@ -71,40 +74,64 @@ trait MesosSchedulerUtils extends Logging {
+      fwInfoBuilder: Protos.FrameworkInfo.Builder): Protos.Credential.Builder = {
+    val credBuilder = Credential.newBuilder()
+    conf.getOption("spark.mesos.principal")
+      .orElse(Option(conf.getenv("SPARK_MESOS_PRINCIPAL")))
```
If you use this environment variable then won't these always be set when using the [Rest client](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/deploy/rest/RestSubmissionClient.scala#L407)? Is this the desired behavior?

rvesse replied:
I am not sure I understand enough of spark internals to answer your question.

These variables are only necessary for Mesos and only on the driver which registers the framework with Mesos. Is it actually possible to submit jobs via REST into a Mesos cluster? Even if it is the Spark framework must exist at that point thereby rendering credentials unnecessary in that scenario.

Or am I missing something here?

ArtRand replied:
Sorry for the delay. I have a use case where I start the Dispatcher in the Mesos cluster and then execute `spark-submit` cluster calls from within the container. Unfortunately this requires me to unset a few environment variables (`MESOS_EXECUTOR_ID MESOS_FRAMEWORK_ID MESOS_SLAVE_ID MESOS_SLAVE_PID MESOS_TASK_ID`) because they interfere with `spark-submit` due to this function in the rest client. 

If the Dispatcher is started in a mode where it needs these Mesos authentication credentials, can we assume that we'll want to always forward them this same way? I realize I might be getting into the weeds here and this might me a _me_ problem. But I thought I'd bring it up.

vanzin replied:
So have you guys reached a conclusion here? I'm not familiar with mesos deployments so I don't really know whether there's a problem here.

ArtRand replied:
I would want to make sure that @susanxhuynh and/or @skonto agree, but I think this is probably fine.

susanxhuynh replied:
I think it's fine to set that env var, SPARK_MESOS_PRINCIPAL.

rvesse committed:
[SPARK-16501] Clarify expected format of files

Note that we expect principal and secret files to be UTF-8 encoded
plaintext

vanzin commented:
ok to test

SparkQA commented:
**[Test build #86946 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86946/testReport)** for PR 20167 at commit [`adc59ea`](https://github.com/apache/spark/commit/adc59ea26c67feff90a259da084b740dd615223f).

SparkQA commented:
**[Test build #86946 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86946/testReport)** for PR 20167 at commit [`adc59ea`](https://github.com/apache/spark/commit/adc59ea26c67feff90a259da084b740dd615223f).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/86946/
Test FAILed.

rvesse committed:
[SPARK-16501] Fix Scala style issues

SparkQA commented:
**[Test build #87103 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87103/testReport)** for PR 20167 at commit [`38d340c`](https://github.com/apache/spark/commit/38d340ce93158d475f1eb7443aa25f65890dc828).

SparkQA commented:
**[Test build #87103 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87103/testReport)** for PR 20167 at commit [`38d340c`](https://github.com/apache/spark/commit/38d340ce93158d475f1eb7443aa25f65890dc828).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87103/
Test FAILed.

rvesse committed:
[SPARK-16501] Fix Scala style issues

SparkQA commented:
**[Test build #87104 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87104/testReport)** for PR 20167 at commit [`8903ba6`](https://github.com/apache/spark/commit/8903ba641b4713799e6739ce71c9baf81183127f).

SparkQA commented:
**[Test build #87104 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87104/testReport)** for PR 20167 at commit [`8903ba6`](https://github.com/apache/spark/commit/8903ba641b4713799e6739ce71c9baf81183127f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87104/
Test PASSed.

rvesse commented:
I've fixed up the Scala style issues so think this is ready to merge

vanzin commented on docs/running-on-mesos.md:
```diff
@@ -82,6 +82,16 @@ a Spark driver program configured to connect to Mesos.
+
+Framework credentials may be specified in a variety of ways depending on your deployment environment and security requirements.  The most simple way is to specify the `spark.mesos.principal` and `spark.mesos.secret` values directly in your Spark configuration.  Alternatively you may specify these values indirectly by instead specifying `spark.mesos.principal.file` and `spark.mesos.secret.file`, t ...
+
+Additionally if you prefer to use environment variables you can specify all of the above via environment variables instead, the environment variable names are simply the configuration settings uppercased with `.` replaced with `_` e.g. `SPARK_MESOS_PRINCIPAL`.  Please note that if you specify multiple ways to obtain the credentials the direct specifications are used in preference to the indirect  ...
```
I think it'd be clearer to specify the order instead of saying "direct vs. indirect". Also because both spark.mesos.principal and SPARK_MESOS_PRINCIPAL to me are direct, and this paragraph doesn't explain which one is chosen.

vanzin commented on docs/running-on-mesos.md:
```diff
@@ -427,15 +437,30 @@ See the [configuration page](configuration.html) for information on Spark config
   <td>(none)</td>
   <td>
-    Set the principal with which Spark framework will use to authenticate with Mesos.
+    Set the principal with which Spark framework will use to authenticate with Mesos.  You can also specify this via the environment variable `SPARK_MESOS_PRINCIPAL`
```
nit: add period

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala:
```diff
@@ -106,6 +99,40 @@ trait MesosSchedulerUtils extends Logging {
+        conf.getOption("spark.mesos.principal.file")
+          .orElse(Option(conf.getenv("SPARK_MESOS_PRINCIPAL_FILE")))
+          .map { principalFile =>
+              Files.toString(new File(principalFile), Charset.forName("UTF-8"))
```
nit: indented too far. Also, we generally use `StandardCharsets.UTF_8`.

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala:
```diff
@@ -106,6 +99,40 @@ trait MesosSchedulerUtils extends Logging {
+      ).foreach { principal =>
+         fwInfoBuilder.setPrincipal(principal)
+         credBuilder.setPrincipal(principal)
+    }
```
nit: indent a little more

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala:
```diff
@@ -106,6 +99,40 @@ trait MesosSchedulerUtils extends Logging {
+        conf.getOption("spark.mesos.secret.file")
+         .orElse(Option(conf.getenv("SPARK_MESOS_SECRET_FILE")))
+         .map { secretFile =>
+           Files.toString(new File(secretFile), Charset.forName("UTF-8"))
```
Same re: `StandardCharsets`.

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala:
```diff
@@ -106,6 +99,40 @@ trait MesosSchedulerUtils extends Logging {
+         }
+      ).foreach { secret =>
+         credBuilder.setSecret(secret)
+    }
```
nit: add indent

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala:
```diff
@@ -106,6 +99,40 @@ trait MesosSchedulerUtils extends Logging {
+           Files.toString(new File(secretFile), Charset.forName("UTF-8"))
+         }
+      ).foreach { secret =>
+         credBuilder.setSecret(secret)
```
nit: this is indented 3 spaces, not 2.

vanzin commented on resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtilsSuite.scala:
```diff
@@ -17,16 +17,21 @@
+import java.io.{File, FileNotFoundException}
+
 import scala.collection.JavaConverters._
+import scala.collection.immutable.Map
```
Why do you need this explicit import?

vanzin commented on resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtilsSuite.scala:
```diff
@@ -237,4 +242,157 @@ class MesosSchedulerUtilsSuite extends SparkFunSuite with Matchers with MockitoS
+    conf.set("spark.mesos.principal.file", "/tmp/does-not-exist")
+
+    intercept[FileNotFoundException] {
+      val credBuilder = utils.buildCredentials(conf, FrameworkInfo.newBuilder())
```
`val credBuilder =` is unnecessary.

vanzin commented on resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtilsSuite.scala:
```diff
@@ -237,4 +242,157 @@ class MesosSchedulerUtilsSuite extends SparkFunSuite with Matchers with MockitoS
+    val conf = new SparkConfWithEnv(Map("SPARK_MESOS_PRINCIPAL_FILE" -> "/tmp/does-not-exist"))
+
+    intercept[FileNotFoundException] {
+      val credBuilder = utils.buildCredentials(conf, FrameworkInfo.newBuilder())
```
`val credBuilder = ` is unnecessary.

vanzin commented on resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtilsSuite.scala:
```diff
@@ -237,4 +242,157 @@ class MesosSchedulerUtilsSuite extends SparkFunSuite with Matchers with MockitoS
+    conf.set("spark.mesos.secret.file", "/tmp/does-not-exist")
+
+    intercept[FileNotFoundException] {
+      val credBuilder = utils.buildCredentials(conf, FrameworkInfo.newBuilder())
```
`val credBuilder = ` is unnecessary.

vanzin reviewed (commented):
Mostly style nits left. Docs could use some clarification.

rvesse committed:
[SPARK-16501] Address PR comments

Style nits and clarify documentation further

rvesse commented:
@vanzin Ok I should hopefully have all those addressed and the documentation clarified appropriately

SparkQA commented:
**[Test build #87258 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87258/testReport)** for PR 20167 at commit [`232043f`](https://github.com/apache/spark/commit/232043f72df53c000efbc789499de05a2208cebe).

rvesse committed:
[SPARK-16501] Add missing punctuation

SparkQA commented:
**[Test build #87259 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87259/testReport)** for PR 20167 at commit [`30e1cbd`](https://github.com/apache/spark/commit/30e1cbd89ca801e0c6d0ac28b5d10554921b72d9).

SparkQA commented:
**[Test build #87258 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87258/testReport)** for PR 20167 at commit [`232043f`](https://github.com/apache/spark/commit/232043f72df53c000efbc789499de05a2208cebe).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87258/
Test PASSed.

SparkQA commented:
**[Test build #87259 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87259/testReport)** for PR 20167 at commit [`30e1cbd`](https://github.com/apache/spark/commit/30e1cbd89ca801e0c6d0ac28b5d10554921b72d9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87259/
Test PASSed.

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala:
```diff
@@ -106,6 +99,40 @@ trait MesosSchedulerUtils extends Logging {
+            Files.toString(new File(principalFile), StandardCharsets.UTF_8)
+          }
+      ).foreach { principal =>
+          fwInfoBuilder.setPrincipal(principal)
```
nit: this block is indented 4 spaces not 2.

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala:
```diff
@@ -106,6 +99,40 @@ trait MesosSchedulerUtils extends Logging {
+      ).foreach { principal =>
+          fwInfoBuilder.setPrincipal(principal)
+          credBuilder.setPrincipal(principal)
+    }
```
nit: this needs to be indented further

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala:
```diff
@@ -106,6 +99,40 @@ trait MesosSchedulerUtils extends Logging {
+           Files.toString(new File(secretFile), StandardCharsets.UTF_8)
+         }
+      ).foreach { secret =>
+          credBuilder.setSecret(secret)
```
same indentation problem here.

vanzin reviewed (commented):
I'll fix the indentation problems before merging, but please try to follow the existing code conventions in the future.

Merging to master.
